### DISCUSSION TITLE
Disable when in a comment?

### DISCUSSION THREAD

**Discussion by ianchesal** [saghen/blink.cmp#564] [Q&A]
I used to do this in `nvim-cmp` with some function logic on the `enable` option like this:

```
local context = require("cmp.config.context")
...stuff....
disabled = disabled or context.in_treesitter_capture("comment")
...stuff...
return not disabled
```

I'm trying to recreate this for blink but I can't seem to detect a comment. I've got this so far:

```
return {
  "saghen/blink.cmp",
  opts = {
    keymap = {
      preset = "default",
    },
    enabled = function()
      local node = vim.treesitter.get_node()
      local disabled = false
      disabled = disabled or (vim.tbl_contains({ "markdown" }, vim.bo.filetype))
      disabled = disabled or (vim.bo.buftype == "prompt")
      disabled = disabled or (node and string.find(node:type(), "comment"))
      return not disabled
    end,
  },
}
```

The:

```
node and string.find(node:type(), "comment")
```

doesn't seem to work correctly. And `enabled` doesn't seem to fire for every treesitter context change.

Is there a way to do this?

I'm using blink via LazyVim if that makes a difference.

**Comment by ianchesal** [comment]
Still struggling to make this happen. [Based on this recipe](https://cmp.saghen.dev/recipes.html#dynamically-picking-providers-by-treesitter-node-filetype) I tried the following but I'm still seeing copilot and dynamic snippet options in the blink suggestions when I'm in a comment block in Ruby or Lua code:

```
return {
  "saghen/blink.cmp",
  opts = {
    enabled = function()
      local disabled = false
      local success, node = pcall(vim.treesitter.get_node)
      -- disabled = disabled or (vim.tbl_contains({ "markdown" }, vim.bo.filetype))
      disabled = disabled or (vim.bo.buftype == "prompt")
      disabled = disabled or (vim.fn.reg_recording() ~= "")
      disabled = disabled or (vim.fn.reg_executing() ~= "")
      disabled = disabled
        or (success and node ~= nil and vim.tbl_contains({ "comment", "line_comment", "block_comment" }, node:type()))

      return not disabled
    end,
  },
}

```

I'm using LazyVim if that matters. Everything is up-to-date.

My full configuration is [here](https://github.com/ianchesal/dotfiles/blob/main/nvim/lua/plugins/blink.lua).

**Reply by bwlee13** [reply]
If you are in a lua comment and you start typing somewhere in the middle of the comment, does this end up working? I can get it working inside a comment, but not from the end of the comment because treesitter.get_node() has it as a table and not a comment or comment_content

**Reply by ianchesal** [reply]
Hmm. I haven't tried that. I'll give it an attempt. My testing has consisted of starting a comment on an empty line.

**Comment by martazvch** [comment]
Interested too

**Comment by bwlee13** [comment]
I DO NOT HAVE THE ANSWER... 

```
return {
  {
    "saghen/blink.cmp",
    dependencies = { "rafamadriz/friendly-snippets" },
    version = "*",
    opts = {
      appearance = {
        use_nvim_cmp_as_default = true,
        nerd_font_variant = "mono",
      },
      keymap = {
        preset = "default",
        ["<C-j>"] = { "select_next", "fallback" },
        ["<C-k>"] = { "select_prev", "fallback" },
        ["<C-u>"] = { "scroll_documentation_up" },
        ["<C-d>"] = { "scroll_documentation_down" },
        ["<C-h>"] = { "hide" },
        ["<CR>"] = { "accept", "fallback" },
      },
      -- this is my test
      -- the following seems to be working
      sources = {
        default = function()
          local node = vim.treesitter.get_node()
          if
            node and vim.tbl_contains({ "comment", "line_comment", "comment_content", "block_comment" }, node:type())
          then
            return { "buffer" }
          else
            return {  "lsp", "path", "snippets", "buffer"  }
          end
        end,
      },

      completion = {
        ghost_text = { enabled = false },
      },
    },
    opts_extend = { "sources.default" },
  },
}

```

**Reply by resistorsoftware** [reply]
I am a bit frustrated by this too. I type Ruby comments and I still get hints when typing them out. Also, i want to leverage this so that when I am typing a string out, like `puts "this is a test string"` that I not get any hints or suggestions in the string. Not sure at all how to pull this off. When I tested your suggestion out, I was still getting hints in my comments.

**Reply by OriLiMu** [reply]
It works for me. Notice to install the parser by command `TSInstall cpp` for example  . you should use ruby I think .

**Reply by resistorsoftware** [reply]
I have Ruby installed. From my debugging it seems that the get_node() function returns the string **method** when in a **comment** and that seems like the likely source of why this is not working for me in Ruby. So now I have to think to myself, why would the Treesitter return **method** instead of **comment** in the get_node function?

**Reply by OriLiMu** [reply]
> I have Ruby installed. From my debugging it seems that the get_node() function returns the string **method** when in a **comment** and that seems like the likely source of why this is not working for me in Ruby. So now I have to think to myself, why would the Treesitter return **method** instead of **comment** in the get_node function?我已经安装了红宝石。从我的调试中，似乎get_node（）函数在**评论**中返回字符串**方法**，这似乎是为什么在Ruby中对我不起作用的可能来源。所以现在我必须对自己思考，为什么TreeTitter返回**方法**而不是在GET_NODE函数中**注释**？

You can try if it works fine in cpp. That you can narrow down the issue to treesitter or ruby parser

**Reply by resistorsoftware** [reply]
It is just a mis-understanding on how treesitter works I guess. the get_node() method from vim does not actually change when editing. I tested on plain old lua as well.

I guess there is some other secret required to turn off completions when inside comments or parenthesis. I will await further help from experts. Clearly my interpretation of how this works is wrong !

**Reply by bwlee13** [reply]
@resistorsoftware welp, this worked for a little bit then stopped working for me and now basically doesn't work at all. I was wrong... get_node() works fine, but at the end of a comment, or even if you start typing a comment, it doesn't think you're in any of the node:type() I have listed in that list. It always thinks you're in a table. Now if you move your cursor to the middle of a comment and start typing, it works. But that definitely isn't the goal.

**Reply by resistorsoftware** [reply]
I spent a fair amount of time trying to figure out how to turn completions on/off based on some criteria, but I have yet to discover the secret sauce. It makes sense to ask Treesitter, hey, what is this current node, but alas, get_node() seems to not be the way.

**Reply by resistorsoftware** [reply]
I was able to get **comment** to show up on occasion, using Ruby as the test. With the `#` symbol. It is not a "condition" in treesitter apparently. Treesitter issues things like *identifier* or *method* as a *Node type* but I can be in a comment, or not, and see the node type of comment, and be in a comment and be typing but the node type is *method*... so this is clearly dumb of me to thing of asking blink to not show completions based on treesitter node type. Yet, that is the recipe! So fun! I have a lot to learn I guess.

**Reply by bwlee13** [reply]
I'm thinking out loud here - what if we set a lua callback code to turn it off based on the line we're on by reading the line and looking for comment syntax? I was thinking about trying this today but it just seems overkill and obviously won't apply for multiline comments. I'm teetering between wanting to get it to work and not wanting to spend hours writing custom code for it haha! But I think a quick function to read the line, ask the lsp what language were on, then checking if this line is a comment.

Any thoughts?

**Reply by resistorsoftware** [reply]
To me it all boils down to Treesitter. I do not know enough about it, but fundamentally, it has the language grammar. So based on that, I cannot believe there is not a way to ask the TS the basic question, what am I inside of? I saw in a Lua file for example, that my node was *inside an if statement* (to paraphrase), so I am surprised we are not getting a *you are inside a comment right now*.

**Reply by bwlee13** [reply]
I know thats where this is tricky. The method I tried at least says "Hey treesitter, where is my cursor?" And it responds with "on a table" or something like that. It works when I'm actually INSIDE a comment, but not at the end of one. There surely is a method for "Hey treesitter, what type of line am i currently on? Is it a comment line??" 
Maybe that's what we need to look for

**Reply by resistorsoftware** [reply]
Ya! I got that gist looking at the table function where you feed in a table with potential comments, and compare it to the current node. And if the node reports *I am a comment* of some sort, bingo. Return false for any kind of completion source. I want the same thing for a "String of text" too, since typing into a string likely needs zero completions. So this is super interesting. I know that recipe for *No Completions Till Special Character is detected!* business, but that seems awfully clunky compared to just asking treesitter, what is this? A comment, a string, a function, a statement, a whatever...

**Reply by bwlee13** [reply]
I was thinking the same thing about the "clunk" of it all

**Comment by DeidaraMC** [comment]
Seems to work good for me simply checking the spot 1 space to the left of the cursor!
You need to put this in the `completion.menu.auto_show` function
```lua
auto_show = function (ctx)
    local row, column = unpack(vim.api.nvim_win_get_cursor(0))
    local success, node = pcall(vim.treesitter.get_node, {
        pos = { row - 1, math.max(0, column - 1) },
        ignore_injections = false
    })
    local reject = { "comment", "line_comment", "block_comment", "string_start", "string_content", "string_end" }
    if success and node and vim.tbl_contains(reject, node:type()) then
        return false;
    end
    -- whatever other logic you want beyond this
    return true
end,
```
I've only tested this fully in Zig

**Reply by resistorsoftware** [reply]
That is a nice addition to the canon here, but as fas as I can tell, it did not work for my version of blink.cmp when I tried it. 

I placed myself inside a string, and suggestions continued to be offered. I worked with comments, and again, suggestions kept being offered. It all seems to be predicated on treesitter and that get_node function which never seems to return the concept of being inside a string or comment very well.

**Reply by DeidaraMC** [reply]
I've also found that removing `buffer` from sources altogether resolves this as well
sources.default = { 'lsp', 'path', 'snippets' }
I can't imagine ever really needing to use the buffer based completions for anything

**Reply by resistorsoftware** [reply]
In plain english, what would a buffer based completion be? If I am working in a file, am I to assume the file is the buffer? So, would a buffer completion not be something along the lines of "Hey, I saw this method somewhere in this file before, therefore you might want to see it again, here, as a completion."

The Lingo of this has not been my strongpoint. So what is the buffer source anyway? I have not found an explanation in plain english for path either. I get LSP and Snippets as sources though... LSP would provide me with some expected built-ins whereas Snippets are custom to say a framework.

**Reply by DeidaraMC** [reply]
I believe the buffer completion just takes words you've already written and suggests them back to you.
So imagine if you're writing in a .txt file there's no lsp or snippets to aid you here, but there are words you've already typed which may or may not be important to you. The buffer completion will give you those.
I turn this off on any IDE I use since I generally only care about what the lsp thinks is important for me (or snippets)

**Reply by resistorsoftware** [reply]
That is what I thought. It does seem to be of very little utility to me then. The idea of removing buffer solutions may then be the ticket!

**Reply by resistorsoftware** [reply]
Unfortunately, even when I remove buffers as a source, snippets kick in at 2 characters and the treesitter problem with nodes still kicks and we get suggestions in a menu while adding a comment or a string. I am guessing the Tree Sitter parser is at fault, but fixing that is a huge task. 

<img width="306" alt="image" src="https://github.com/user-attachments/assets/2f1b9822-1c38-455b-a301-157b7d9ee3f2" />

**Comment by jtackaberry** [comment]
I was migrating from AstroNvim v4 to v5 which, among other things, has replaced nvim-cmp with blink.nvim.  With every other customization due to a replaced plugin that I had to reimplement, I'd never have imagined _this_ would be the thing to do me in.

Despite various suggestions in this discussion, the only thing that actually worked for me was ripping the logic out of nvim-cmp and using that.  nvim-cmp can also handle non-treesitter cases, but for my purposes I was fine with requiring TS.

```lua
-- Lifted from https://github.com/hrsh7th/nvim-cmp/blob/v0.0.2/lua/cmp/config/context.lua
function in_treesitter_capture(capture)
  local row, col = unpack(vim.api.nvim_win_get_cursor(0))
  if vim.api.nvim_get_mode().mode == 'i' then
    col = col - 1
  end

  local buf = vim.api.nvim_get_current_buf()
  local get_captures_at_pos = require('vim.treesitter').get_captures_at_pos

  local captures_at_cursor = vim.tbl_map(function(x)
    return x.capture
  end, get_captures_at_pos(buf, row - 1, col))

  if vim.tbl_isempty(captures_at_cursor) then
    return false
  elseif type(capture) == 'string' and vim.tbl_contains(captures_at_cursor, capture) then
    return true
  elseif type(capture) == 'table' then
    for _, v in ipairs(capture) do
      if vim.tbl_contains(captures_at_cursor, v) then
        return true
      end
    end
  end
  return false
end

return {
  "Saghen/blink.cmp",
  opts = {
    completion = {
      menu = {
        auto_show = function()
          return not in_treesitter_capture("comment") and not require("luasnip").expand_or_jumpable()
        end
      }
    }
  },
}
```

**Comment by becknik** [comment]
This is what I'm using in Neovim 0.11.2 to enable git commit & emoji cmp with blink in comments. This surely is a bit off-topic, but maybe it might provide some help.

I basically took the code from @DeidaraMC and added the {buffer} property:

```lua
function is_in_string_like_node()
  if vim.bo.buftype == "prompt" then
    return false
  elseif vim.tbl_contains({ 'gitcommit', 'markdown' }, vim.bo.filetype) then
    return true
  end

  local row, column = unpack(vim.api.nvim_win_get_cursor(0))
  local success, node = pcall(vim.treesitter.get_node, {
    bufnr = 0,
    pos = { row - 1, math.max(0, column - 1) } -- seems to be necessary...
  })

  -- `string_fragment` is a node type reported in react files, so I've added it
  local reject = { "comment", "line_comment", "block_comment", "string_start", "string_fragment", "string_content", "string_end" }
  return (success and node and vim.tbl_contains(reject, node:type()))
end
```

I'm using this function reference in the table path `providers.(git|emoji).should_show_items`

**Comment by beetol** [comment]
Anyone who is looking for an answer, here is what worked for me.

```
local function in_comment()
  local row, col = unpack(vim.api.nvim_win_get_cursor(0))
  local parser = vim.treesitter.get_parser(0)
  if not parser then return false end
  parser:parse()
  local node = vim.treesitter.get_node({ pos = { row - 1, math.max(0, col - 1) } })
  if not node then return false end
  while node do
    if node:type() == 'comment' then return true end
    node = node:parent()
  end
  return false
end

return {
  'saghen/blink.cmp',
  opts = {
    enabled = function()
      local ft = vim.bo.filetype
      if ft == 'markdown' or ft == 'text' then return false end
      return not in_comment()
    end,
    completion = {
      list = {
        max_items = 10,
      },
      trigger = {
        show_in_snippet = false,
      },
      menu = {
        auto_show = false,
        border = 'rounded',
      },
      documentation = {
        auto_show = true,
        window = { border = 'rounded' },
      },
      ghost_text = {
        enabled = function() return not in_comment() end,
        show_without_selection = true,
        show_with_menu = false,
      },
    },
}
```

This is a gist of my config so I'm highlighting the parts that worked for me. I use a keymapping to bring up the completion window but mostly rely on ghost_text to complete suggestions. The problem I had was typing within a comment and blink's ghost text completion would still pop up, which meant the completion window was still possible to bring up with my keymap.
